### ISSUE TITLE
Decide How to handle `Transaction Submission Timeout`

### ISSUE THREAD

**Issue by b-yap** [pendulum-chain/spacewalk#558]
From the [failed CI](https://github.com/pendulum-chain/spacewalk/actions/runs/11176052005/job/31071234041?pr=545#step:4:5250):
> ERROR transfer_stellar_asset{request_type=Redeem request_id=0xb4e5d0da87b0b1ad3cf0ed0295721d23be3dce617cb6667c5c41afbf82df62fb}: wallet::horizon::responses: Response returned an error: HorizonSubmissionError { title: "Transaction Submission Timeout", status: 504, reason: "Your transaction submission request has timed out. This does not necessarily mean the submission has failed. Before resubmitting, please use the transaction hash provided in `extras.hash` to poll the GET /transactions endpoint for sometime and check if it was included in a ledger.", result_code_op: [], envelope_xdr: None }

After returning the same transaction, there must have been race conditions, as the next error that appeared:
> ERROR transfer_stellar_asset{request_type=Redeem request_id=0xb4e5d0da87b0b1ad3cf0ed0295721d23be3dce617cb6667c5c41afbf82df62fb}: wallet::horizon::responses: Response returned an error: HorizonSubmissionError { title: "Transaction Failed", status: 400, reason: "tx_bad_seq", result_code_op: [], envelope_xdr: Some("AAAAAgAAAABIAqJ3gBZp7ZPNNB6lqFRtrHwsVORcDOsAk9uqkBeoEwAAAGQAACygAAABHAAAAAAAAAABAAAAHERCOWR0ODFLZUxzVUhOVzRFU0dCTHlCNHl6akIAAAABAAAAAQAAAABIAqJ3gBZp7ZPNNB6lqFRtrHwsVORcDOsAk9uqkBeoEwAAAAEAAAAAvLWEHHR8B5HLUxEsvLtwpqTLhqOB93KKa2mitHztRHAAAAABVVNEQwAAAAAU0ZYxsDcX2auaNm4QMh7iZucux2yrYZDwoTNtSCKfiwAAAAAAABODAAAAAAAAAAGQF6gTAAAAQAgHmNMxHyjTF1s4R4vIieggkkHbHCYY69Ccgd+TwouhD0laCxZOhF+MJ6UiF5XdWWoaPk75PAXw7tBKOc6Skw0=") }

### Discuss how to handle this:
* follow the instructions about polling  `/transactions`;
* do not perform resubmission for `Transaction Submission Timeout`s; OR
* stay as is; no code changes

**Comment by ebma** [comment]
Let's align on what happened here. The first submission timed out and the second submission for the _same_ transaction (same sequence number) returned a `tx_bad_seq` as the first one already got accepted. 

Can you link to the relevant parts of the code that handle this case? Don't we already check if a transaction was submitted in the past by going through all historic transactions of an account? Or not in this case? 

I think the best-practice way to deal with these 504 submission errors is to try and resubmit until it gets accepted.

**Comment by b-yap** [comment]
> Don't we already check if a transaction was submitted in the past by going through all historic transactions of an account

Not when we're submitting it on-the-go; meaning it's **_NOT_** from the periodic resubmission. (It won't pass through any logic in `clients/wallet/src/resubmissions.rs`).

> I think the best-practice way to deal with these 504 submission errors returned by Stellar horizon servers is to try and resubmit until it gets accepted.

We are. There's a delay in-between though. Probably because of it, another  (successful) transaction took the same sequence number. Once resubmission is triggered, the sequence number is already invalid.

Also, this might have happen because the 2 CI tests ran on parallel: 
* Run Tests for Pallets and Clients
* Run Tests for Vault
Although it's hard to trace this. I searched for sequence _49065706389788_; it onlu appeared in the job _Run Tests for Vault_

I've added the links of the relevant code in the description.

**Comment by b-yap** [comment]
This has only happened one time; I  would like to observe if it happens again.
And if it does, I think one non-drastic solution is to reduce the delay. Right now it's set to 10 seconds MULTIPLIED by the delay counter:
https://github.com/pendulum-chain/spacewalk/blob/7c7989875e95e1cfe3b0aeba9bb6af01d9e33e58/clients/wallet/src/horizon/horizon.rs#L176

Default value of max_backoff_delay is:
https://github.com/pendulum-chain/spacewalk/blob/7c7989875e95e1cfe3b0aeba9bb6af01d9e33e58/clients/wallet/src/stellar_wallet.rs#L67
which is huge.

**Comment by ebma** [comment]
You are right, that's probably a bit too much and I'm also fine if we reduce it. `10` sounds good to me. This makes the error you describe in the ticket more unlikely to occur but it could still happen. Correct me if I'm wrong but when this happens, the vault will not attempt to retry the execution of the redeem request until it restarts, correct?

**Comment by b-yap** [comment]
> the vault will not attempt to retry the execution of the redeem request until it restarts, correct?

The envelope will be removed (be it successful or not):
https://github.com/pendulum-chain/spacewalk/blob/7c7989875e95e1cfe3b0aeba9bb6af01d9e33e58/clients/wallet/src/stellar_wallet.rs#L282

**Comment by ebma** [comment]
Right but when the vault restarts, it will probably see the redeem request still pending and then retry the whole redeem process again. In this case, this is also not problematic except that for the longer waiting time of the user for the redeem. 

However, let's say your assumption is correct and this occurred because 2 CIs were running in parallel, then we are very unlikely to encounter the same issue in production as long as only one vault client is running per vault (and the Stellar account of the vault is not re-used twice) which is something we explicitly mention in the vault operator documentation. 

So let's agree that we change the default max backoff and keep the rest as is, or what do you think @b-yap?

**Comment by prayagd** [comment]
@ebma can you please share a brief about what is the problem and should we prioritise it?

**Comment by ebma** [comment]
I changed the description of the ticket to summarize the conclusion based on the comments. We don't need to prioritize it as it's an edge case that should not happen in practice.
